src/Processor/ForeachProcessor.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+namespace Elastica\Processor;
+
+/**
+ * Elastica Foreach Processor.
+ *
+ * @author Federico Panini <fpanini@gmail.com>
+ * @author Thibaut Simon-Fine <tsimonfine@gmail.com>
+ *
+ * @see https://www.elastic.co/guide/en/elasticsearch/reference/current/foreach-processor.html
+ */
+class ForeachProcessor extends AbstractProcessor
+{
+    public const DEFAULT_IGNORE_MISSING_VALUE = false;
+    protected const PROCESSOR_NAME = 'foreach';
+
+    /**
+     * @param AbstractProcessor|array $processor
+     */
+    public function __construct(string $field, $processor)
+    {
+        $this->setField($field);
+
+        if ($processor instanceof AbstractProcessor) {
+            $this->setProcessor($processor);
+        } elseif (\is_array($processor)) {
+            $this->setRawProcessor($processor);
+        } else {
+            throw new \TypeError('Expected Elastica\Processor\AbstractProcessor or array');
+        }
+    }
+
+    /**
+     * Set field.
+     *
+     * @return $this
+     */
+    public function setField(string $field): self
+    {
+        return $this->setParam('field', $field);
+    }
+
+    /**
+     * Set processor.
+     *
+     * @param AbstractProcessor
+     *
+     * @return $this
+     */
+    public function setProcessor(?AbstractProcessor $processor): self
+    {
+        return $this->setParam('processor', $processor);
+    }
+
+    /**
+     * Set raw processor.
+     * Example : ['remove' => ['field' => 'user_agent']].
+     *
+     * @return $this
+     */
+    public function setRawProcessor(array $processor): self
+    {
+        return $this->setParam('processor', $processor);
+    }
+
+    /**
+     * Set ignore_missing. Default value false.
+     *
+     * If true and field does not exist or is null, the processor quietly exits without modifying the document
+     *
+     * @return $this
+     */
+    public function setIgnoreMissing(bool $ignoreMissing): self
+    {
+        return $this->setParam('ignore_missing', $ignoreMissing);
+    }
+
+    /**
+     * Param's name
+     * Picks the last part of the class name and makes it snake_case
+     * You can override this method if you want to change the name.
+     *
+     * @return string name
+     */
+    protected function _getBaseName()
+    {
+        return self::PROCESSOR_NAME;
+    }
```

Comment:
To avoid this, maybe the class should be named `ForEach`?

Reply:
"foreach" is a php construct, don't think it's a good idea.

And my PHP code check doesn't like that.
```
syntax error, unexpected 'ForEach' (T_FOREACH), expecting identifier (T_STRING)
Unexpected 'foreach'. Expected 'Name'.intelephense(1001)
Expression is not writable.intelephense(1024)
```

Reply:
Sorry @ThibautSF I meant `ForEachProcessor`, yes `Foreach` is a reserved keyword.

Reply:
`Elastica\Processor\ForeachProcessor._getBaseName()` is an override of default `Elastica\Param._getBaseName()` which only calls and return the static method `Elastica\Utils.getParamName($class)`

```php
    /**
     * Tries to guess the name of the param, based on its class
     * Example: \Elastica\Query\MatchAll => match_all.
     *
     * @param object|string Object or class name
     * @param mixed $class
     *
     * @return string parameter name
     */
    public static function getParamName($class)
    {
        if (\is_object($class)) {
            $class = \get_class($class);
        }

        $parts = \explode('\\', $class);
        $last = \array_pop($parts);
        $last = \preg_replace('/Query$/', '', $last); // for BoolQuery and MatchQuery

        return self::toSnakeCase($last);
    }
```

As we can see, this method with "ForEachProcessor" class may return "for_each_processor".
Note: When I tried "ForeachProcessor" before adding the method override, the generated processor for elasticsearch was "foreach_processor" thus I pushed https://github.com/ruflin/Elastica/pull/1813/commits/e7659be9d92078ad97a48bd994f580eaa01abb1a
> Fix processor name being "foreach_processor" instead of "processor" due to class name (note: 'Foreach' is reserved)

So, the override is there to ensure that the param name used will be "foreach".
Note: The presence of the const aims to simplify processor name modification (idk, in case Elasticsearch dev wants to change "foreach" with "for")

Reply:
Thanks for your detailed explanation, I better understand now ;)

What about in this case adding:
```php
abstract class AbstractProcessor extends Param
{
    protected function _getBaseName()
    {
        $shortName = (new \ReflectionClass($this))->getShortName();
        $shortName = \preg_replace('/Processor$/', '', $shortName);

        return Util::toSnakeCase($shortName);
    }
}
```

Or add `\preg_replace('/Processor$/', '', $last);` in `Util::getParamName()` but impact can be bigger.

For fun I benchmarked the use of `ReflectionClass` vs `explode` to see if it impacts on performance ;)
See https://3v4l.org/9rlBF

Reply:
> What about in this case adding:
> 
> ```
> abstract class AbstractProcessor extends Param
> {
>     protected function _getBaseName()
>     {
>         $shortName = (new \ReflectionClass($this))->getShortName();
>         $shortName = \preg_replace('/Processor$/', '', $shortName);
> 
>         return Util::toSnakeCase($shortName);
>     }
> }
> ```
> 
> Or add `\preg_replace('/Processor$/', '', $last);` in `Util::getParamName()` but impact can be bigger.

I don't know... the current implementation is working for every case (generic) and was designed to allow custom name (adaptability).

In my opinion, we need to keep in mind that the issue ("foreach" being reserved) which forces us to use the non-generic way is uncommon. And if tomorrow we need a class "FooProcessor" which need to be interpreted by _getBaseName() as "foo_processor" the `preg_replace('/Processor$/', '', $shortName);` might be annoying.


> For fun I benchmarked the use of `ReflectionClass` vs `explode` to see if it impacts on performance ;)
> See https://3v4l.org/9rlBF

You might be interested to look at this comment https://www.php.net/manual/fr/function.get-class.php#114568
It's 6 years old so might need to benchmark again, but `strrpos()` looks like better than `explode()`